mpc/mpc-core/src/commit.rs

```diff
@@ -0,0 +1,124 @@
+//! This module provides a hash commitment scheme for types which implement `serde::Serialize`
+
+use crate::hash::{Hash, SecureHash};
+use rand::{thread_rng, Rng};
+use serde::{Deserialize, Serialize};
+
+/// Error associated with commitments
+#[derive(Debug, thiserror::Error)]
+#[allow(missing_docs)]
+pub enum CommitmentError {
+    #[error("Invalid decommitment")]
+    InvalidDecommitment,
+}
+
+/// A randomly generated 32 byte nonce
+#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
+pub struct Nonce([u8; 32]);
+
+impl Nonce {
+    /// Creates a random 32 byte nonce
+    pub fn random() -> Self {
+        Self(thread_rng().gen())
+    }
+}
+
+/// Decommitment data for a commitment
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Decommitment<T>
+where
+    T: Serialize,
+{
+    nonce: Nonce,
+    data: T,
+}
+
+impl<T> Decommitment<T>
+where
+    T: Serialize,
+{
+    /// Creates a new decommitment
+    pub fn new(data: T) -> Self {
+        Self {
+            nonce: Nonce::random(),
+            data,
+        }
+    }
+
+    /// Creates a hash commitment
+    pub fn commit(&self) -> Hash {
```

Comment:
Can we rename this, or maybe the other one, as it collides with `HashCommit::commit`?